File: docs/js/imageopt.js
// Canvas-based image processing — ported from mangaconv/core/imageopt.py.
// Runs entirely in the browser: resize to native resolution, grayscale,
// autocontrast, adaptive JPEG quality, and spread splitting.

// Decode bytes into an ImageBitmap (fast, off-thread decode).
async function decode(bytes) {
  const blob = new Blob([bytes]);
  return await createImageBitmap(blob);
}

// Scale to fit within width x height preserving aspect ratio (never distorts).
function fitDimensions(srcW, srcH, width, height) {
  const targetRatio = width / height;
  const srcRatio = srcW / srcH;
  if (srcRatio > targetRatio) {
    return [width, Math.max(1, Math.round(width / srcRatio))];
  }
  return [Math.max(1, Math.round(height * srcRatio)), height];
}

function newCanvas(w, h) {
  if (typeof OffscreenCanvas !== "undefined") return new OffscreenCanvas(w, h);
  const c = document.createElement("canvas");
  c.width = w;
  c.height = h;
  return c;
}

// Histogram-based autocontrast equivalent to PIL autocontrast(cutoff=1):
// clip 1% of pixels from each end of the luminance histogram, then stretch.
function applyAutocontrast(data, cutoffPercent = 1) {
  const hist = new Array(256).fill(0);
  const n = data.length / 4;
  for (let i = 0; i < data.length; i += 4) hist[data[i]]++; // R==G==B after grayscale
  const cut = Math.floor((n * cutoffPercent) / 100);

  let lo = 0;
  let acc = 0;
  for (; lo < 256; lo++) {
    acc += hist[lo];
    if (acc > cut) break;
  }
  let hi = 255;
  acc = 0;
  for (; hi > 0; hi--) {
    acc += hist[hi];
    if (acc > cut) break;
  }
  if (hi <= lo) return; // degenerate; leave as-is

  const scale = 255 / (hi - lo);
  const lut = new Uint8ClampedArray(256);
  for (let v = 0; v < 256; v++) {
    lut[v] = Math.max(0, Math.min(255, Math.round((v - lo) * scale)));
  }
  for (let i = 0; i < data.length; i += 4) {
    const v = lut[data[i]];
    data[i] = data[i + 1] = data[i + 2] = v;
  }
}

// Encode a canvas to a JPEG/PNG Blob at the given quality.
async function encode(canvas, { type, quality }) {
  if (canvas.convertToBlob) return await canvas.convertToBlob({ type, quality });
  return await new Promise((res) => canvas.toBlob(res, type, quality));
}

// Process one bitmap (already cropped if it was a spread half) into final bytes.
async function processBitmap(bitmap, width, height, opts) {
  const {
    grayscale = true,
    autocontrast = true,
    maxQuality = 0.9,
    minQuality = 0.4,
    targetBytes = null,
    preferPng = false,
  } = opts;

  const [w, h] = fitDimensions(bitmap.width, bitmap.height, width, height);
  const canvas = newCanvas(w, h);
  const ctx = canvas.getContext("2d", { willReadFrequently: true });
  ctx.imageSmoothingEnabled = true;
  ctx.imageSmoothingQuality = "high";
  // White background so transparency flattens to white (manga pages are white).
  ctx.fillStyle = "#ffffff";
  ctx.fillRect(0, 0, w, h);
  ctx.drawImage(bitmap, 0, 0, w, h);

  if (grayscale || autocontrast) {
    const img = ctx.getImageData(0, 0, w, h);
    const d = img.data;
    if (grayscale) {
      for (let i = 0; i < d.length; i += 4) {
        // Rec. 601 luma, matching PIL's "L" conversion.
        const v = (d[i] * 299 + d[i + 1] * 587 + d[i + 2] * 114 + 500) / 1000 | 0;
        d[i] = d[i + 1] = d[i + 2] = v;
      }
    }
    if (autocontrast) applyAutocontrast(d, 1);
    ctx.putImageData(img, 0, 0);
  }

  if (preferPng) {
    const blob = await encode(canvas, { type: "image/png" });
    return { blob, width: w, height: h, mediaType: "image/png", ext: "png" };
  }

  if (targetBytes == null) {
    const blob = await encode(canvas, { type: "image/jpeg", quality: maxQuality });
    return { blob, width: w, height: h, mediaType: "image/jpeg", ext: "jpg" };
  }

  // Adaptive quality: largest q in [min,max] whose size <= target (binary search
  // over a quantized 0..100 scale to keep the number of encodes small).
  let best = await encode(canvas, { type: "image/jpeg", quality: minQuality });
  let lo = Math.round(minQuality * 100);
  let hi = Math.round(maxQuality * 100);
  while (lo <= hi) {
    const mid = (lo + hi) >> 1;
    const cand = await encode(canvas, { type: "image/jpeg", quality: mid / 100 });
    if (cand.size <= targetBytes) {
      best = cand;
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }
  return { blob: best, width: w, height: h, mediaType: "image/jpeg", ext: "jpg" };
}

// Crop a landscape bitmap into two portrait halves (RTL: right half first).
function spreadHalves(bitmap, rightToLeft) {
  const mid = Math.floor(bitmap.width / 2);
  const left = { x: 0, w: mid };
  const right = { x: mid, w: bitmap.width - mid };
  return rightToLeft ? [right, left] : [left, right];
}

async function cropBitmap(bitmap, region) {
  return await createImageBitmap(bitmap, region.x, 0, region.w, bitmap.height);
}

// Public: process raw page bytes, optionally splitting detected spreads.
// Returns an array of { blob, width, height, mediaType, ext }.
export async function processPage(bytes, width, height, opts = {}) {
  const { splitSpreads = false, rightToLeft = true } = opts;
  const bitmap = await decode(bytes);
  try {
    if (splitSpreads && bitmap.width > bitmap.height) {
      const out = [];
      for (const region of spreadHalves(bitmap, rightToLeft)) {
        const half = await cropBitmap(bitmap, region);
        out.push(await processBitmap(half, width, height, opts));
        half.close?.();
      }
      return out;
    }
    return [await processBitmap(bitmap, width, height, opts)];
  } finally {
    bitmap.close?.();
  }
}

File: mangaconv/core/imageopt.py
"""Image processing tuned for e-ink readers.

Key findings (see SPEC.md):

* Resize to the device's *exact native resolution*. Downscaling "in the dark"
  (e.g. an arbitrary 1050px) is what makes text look bad. Fit the page inside
  the native panel, preserving aspect ratio.
* E-ink is grayscale, so converting to mode ``L`` (8-bit) saves a lot of bytes
  with no visible loss.
* ``ImageOps.autocontrast(cutoff=1)`` noticeably improves perceived contrast.
* JPEG quality should be chosen adaptively: pick the highest ``q`` (<= cap)
  that keeps the encoded page under budget.
"""

from __future__ import annotations

import io
from dataclasses import dataclass

from PIL import Image, ImageOps

# Be tolerant of slightly broken JPEGs that appear in real archives.
ImageFile_LOAD_TRUNCATED = True
try:  # pragma: no cover - defensive
    from PIL import ImageFile

    ImageFile.LOAD_TRUNCATED_IMAGES = True
except Exception:  # pragma: no cover
    pass


@dataclass
class ProcessedPage:
    """A page after processing, ready to embed in an EPUB."""

    data: bytes
    width: int
    height: int
    media_type: str  # "image/jpeg" or "image/png"
    ext: str  # "jpg" or "png"


def _load(data: bytes) -> Image.Image:
    img = Image.open(io.BytesIO(data))
    img.load()
    return img


def split_spread(img: Image.Image, *, right_to_left: bool = True) -> list[Image.Image]:
    """Split a landscape (double-page) image into two portrait halves.

    For manga (right-to-left), the right half is page N and the left half is
    page N+1, so the right half is yielded first.
    """
    w, h = img.size
    mid = w // 2
    left = img.crop((0, 0, mid, h))
    right = img.crop((mid, 0, w, h))
    return [right, left] if right_to_left else [left, right]


def fit_to_screen(img: Image.Image, width: int, height: int) -> Image.Image:
    """Scale ``img`` to fit within ``width x height`` preserving aspect ratio.

    Uses high-quality Lanczos resampling. The result is never upscaled beyond
    the native panel; it is scaled so the longest dimension matches the screen.
    """
    target_ratio = width / height
    src_ratio = img.width / img.height
    if src_ratio > target_ratio:
        new_w = width
        new_h = max(1, round(width / src_ratio))
    else:
        new_h = height
        new_w = max(1, round(height * src_ratio))
    return img.resize((new_w, new_h), Image.LANCZOS)


def _encode(img: Image.Image, *, quality: int, fmt: str) -> bytes:
    buf = io.BytesIO()
    if fmt == "PNG":
        img.save(buf, format="PNG", optimize=True)
    else:
        img.save(buf, format="JPEG", quality=quality, optimize=True, progressive=True)
    return buf.getvalue()


def process_image(
    img: Image.Image,
    width: int,
    height: int,
    *,
    grayscale: bool = True,
    autocontrast: bool = True,
    max_quality: int = 90,
    min_quality: int = 40,
    target_bytes: int | None = None,
    prefer_png: bool = False,
) -> ProcessedPage:
    """Process one already-decoded image into final bytes.

    If ``target_bytes`` is given, performs a binary search for the highest
    JPEG quality (<= ``max_quality``) that fits the budget, never dropping
    below ``min_quality``.
    """
    # Flatten transparency onto white before mode conversion.
    if img.mode in ("RGBA", "LA", "P"):
        img = img.convert("RGBA")
        bg = Image.new("RGBA", img.size, (255, 255, 255, 255))
        img = Image.alpha_composite(bg, img).convert("RGB")

    if grayscale:
        img = img.convert("L")
    elif img.mode not in ("RGB", "L"):
        img = img.convert("RGB")

    img = fit_to_screen(img, width, height)

    if autocontrast:
        img = ImageOps.autocontrast(img, cutoff=1)

    fmt = "PNG" if prefer_png else "JPEG"
    if fmt == "PNG":
        data = _encode(img, quality=max_quality, fmt="PNG")
        return ProcessedPage(data, img.width, img.height, "image/png", "png")

    if target_bytes is None:
        data = _encode(img, quality=max_quality, fmt="JPEG")
        return ProcessedPage(data, img.width, img.height, "image/jpeg", "jpg")

    # Adaptive quality: largest q in [min, max] with size <= target.
    best = _encode(img, quality=min_quality, fmt="JPEG")
    lo, hi = min_quality, max_quality
    while lo <= hi:
        mid = (lo + hi) // 2
        candidate = _encode(img, quality=mid, fmt="JPEG")
        if len(candidate) <= target_bytes:
            best = candidate
            lo = mid + 1
        else:
            hi = mid - 1
    return ProcessedPage(best, img.width, img.height, "image/jpeg", "jpg")


def process_page(
    data: bytes,
    width: int,
    height: int,
    *,
    grayscale: bool = True,
    autocontrast: bool = True,
    max_quality: int = 90,
    split_spreads: bool = False,
    right_to_left: bool = True,
    **kwargs,
) -> list[ProcessedPage]:
    """Process raw page bytes, optionally splitting detected spreads.

    Returns a list because one input spread can yield two output pages.
    A page is considered a spread when it is landscape (wider than tall).
    """
    img = _load(data)
    if split_spreads and img.width > img.height:
        halves = split_spread(img, right_to_left=right_to_left)
        return [
            process_image(
                half, width, height,
                grayscale=grayscale, autocontrast=autocontrast,
                max_quality=max_quality, **kwargs,
            )
            for half in halves
        ]
    return [
        process_image(
            img, width, height,
            grayscale=grayscale, autocontrast=autocontrast,
            max_quality=max_quality, **kwargs,
        )
    ]
